### ISSUE TITLE
[New Concept Exercise] 

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2370]
This issue describes how to implement  the  `context-manager customization` **concept exercise** for the Python track.
The related **concept documents issue** can be found [here](https://github.com/exercism/python/issues/3081).

<br>

## ✅  Getting started

_If you have not yet created or contributed to a concept exercise, this issue will require some upfront reading to give you the needed background knowledge. Some good example exercises to look at in the repo:_

<details>
<summary>💡<b>Example Exercises</b>💡</summary>
<br>

   1. [Little Sister's Vocabulary](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-vocab)
   2. [Meltdown Mitigation](https://github.com/exercism/python/tree/main/exercises/concept/meltdown-mitigation) 
   3. [Making the Grade](https://github.com/exercism/python/tree/main/exercises/concept/making-the-grade)
   4. [Ellen's Alien Game](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game)

</details>

We also recommend completing one or more of the concept exercises (_they're called "learning exercises"_) [on the website](https://exercism.org/tracks/python/concepts). 

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please go through the following documents:

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concept Exercises**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concept Exercises?](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)

## 🎯   Goal

The goal of the concept exercise described in this issue is to teach `context- manager customization` (_the machinery behind the with keyword_) in Python.

<br>

## 💡  Learning objectives

- Review what the `with` keyword *is* and what it _does_ in Python
  - context-handling & scope
  - setup and teardown processes
  - memory and resource allocation
- Understand different `context-manager` Types and their uses
- Get familiar with some of the  utilities & decorators in `contextlib`, and how they're used to customize a `context-manager`
  - `@contextlib.contextmanager`
  - `contextlib.closing()`
  - `contextlib.nullcontext()`
  - `contextlib.suppress`
  - `conextlib.redirect_stdout()`
  - `contextlib.redirect_stderr()`
  - `class contextlib.ContextDecorator`
- Understand how customized `context-managers` are used with the `with` statement
    - `__enter__`
    - `__exit__` 
- Implement some of the most common use cases for managing / customizing management of resources with the `with` keyword (_maybe some of the recipes listed in the [docs](https://docs.python.org/3/library/contextlib.html#examples-and-recipes)?_)
  - file opening/closing/processing
  - database connectivity
  - error handling
  - logging
  - testing

<br>

## 🚫  Topics that are Out of scope

While many of these topics could be considered prerequisites or necessary building blocks for this concept, we've marked them as "out of scope".  

This means that we'd like you to focus instruction and examples  primarily on the topic of **this** concept (_context managers and customization_), and save any specific instruction or custom use of the topics below for exercises that focus on those concepts in detail.  

**Some examples:**  `@contextmanager` requires a `generator` to act as a factory.  But this exercise should not teach the making of or customization of a `generator`, since `generators` have their own exercise.  The same goes for creating `classes`,  `raising` & `handling` errors, or using `assert` for testing.  The overall goal should be to use _just enough_ to explain/instruct the student in the `context-manager` customization concept, and assume they have learned or will learn the other details.  If you have additional questions, **please reach out to one of the maintainers.**

<br>
<details>
<summary><b>Out of Scope Topics for Context Manager Customization</b></summary>
<br>

- `async` versions of `context-managers` (_we are leaving `async-io` and `async` off our concept list right now due to their complexity_)
- `classes`  & `class customization`
- `class-inheritance` 
- `comprehensions` 
- `comprehensions` in `lambdas`
- `coroutiens` 
- `decorators` 
- `error-handling` 
- `raising errors`
- `functions` and `higher-order functions`  
-  `functools` and related `map()`, `filter()` and `functools.reduce()`
- `generators`
- `lambdas` 
- using an `assignment expression` or "walrus" operator (`:=`) 
- `enums`
- `testing` and `assert`

</details>
<br>

## 🤔  Concepts

- `with`
- `context-managers`
- `contextlib`
-  Special Methods
   -   `__enter__`
   -  `__exit__`

<br>

## ↩️  Prerequisites

<details>
<summary>Prereqs</summary>

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

- `basics`
- `bools`
- `classes`  
- `class customization` 
- `class-inheritance` 
- `comparisons`
- `rich-comparisons`
- `decorators`
- `dicts`
- `dict-methods`
- `raising-and-handling-errors` 
- `functions`
- `generators`
- `higher-order-functions`
- `lists`
- `list-methods`
- `loops`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `testing`
- `tuples`
- `with-statement`

</details>

<br>

## 📚  Resources for Writing & Reference

<details>
<summary><b>Resources</b></summary>

- [Python Docs: Compound Statements](https://docs.python.org/3/reference/compound_stmts.html)
- [Python Docs:  With Statement Context Managers](https://docs.python.org/3/reference/datamodel.html#with-statement-context-managers)
- [Python Docs: With Statement](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)
  - [Python Docs: Reading and Writing Files](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files)
- [Python Docs:  Context Manager Types](https://docs.python.org/3/library/stdtypes.html#context-manager-types)
- [PEP 0343: The `with` Statement](https://www.python.org/dev/peps/pep-0343/)
- [PEP 0343:  Examples](https://www.python.org/dev/peps/pep-0343/#examples)
- [Python Docs: `contextlib`](https://docs.python.org/3/library/contextlib.html#module-contextlib)
- [Python Module of the Week: `contextlib`](https://pymotw.com/3/contextlib/)
- [Python Morsels:  Context-managers](https://www.pythonmorsels.com/topics/context-managers/)
-  [Dan Bader: Context Managers and the `with` statement in Python](https://dbader.org/blog/python-context-managers-and-with-statement)
- [Real Python: Context managers](https://realpython.com/python-with-statement/#:~:text=Coding%20Class%2DBased%20Context%20Managers,-To%20implement%20the%20context%20management)
- [Innoplex: Fun With Context Managers](https://www.innoplexllc.com/post/python-fun-with-context-managers)
- [Siv Scripts:  A Gentle Introduction to Context Managers](https://alysivji.github.io/managing-resources-with-context-managers-pythonic.html)
- [arnavk.org: Context Managers in the Real World](http://arnavk.org/posts/python-context-managers/)
- [Abhinav Gupta:  Go Antipatterns - With* Context Managers](https://abhinavg.net/posts/context-manager-antipattern/)

</details>
<details>
<summary><b>Additional Articles</b></summary>

-  [David Beazley at (Unenix.org) Python: With That - 5 Easy Context Managers ](https://www.usenix.org/system/files/login/articles/12_beazley-online.pdf)
- [Jeff Knupp: Python with Context Managers](https://jeffknupp.com/blog/2016/03/07/python-with-context-managers/)
-  [Red's Digressions: The Curious Case of Python's Context Manager](https://rednafi.github.io/digressions/python/2020/03/26/python-contextmanager.html)
- [Cambridge Spark: Python Context manager](https://blog.cambridgespark.com/python-context-manager-3d53a4d6f017)

</details>
<br>  

## 📁  Exercise Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This can summarize/paraphrase the linked concept documents if they have already been created (either the `about` or the `introduction`).   The summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.
 
* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise `Exemplar.py` Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and **does not use single letter variable names**.  It should also include proper module and function-level docstrings.  However, it should **NOT** include typehinting or type aliases.

* ### `<Exercise>.py` (Stub) for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### `<Exercise>_Test.py` Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise `Hints.md`

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.
 
* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>
</td>
</table>
</details>
<br>

## ♾️  Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  `a59b4983-29b3-4519-bb86-eb3de45c89b0`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue    

<br>

## 🎶 Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## 🆘   Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by DjangoFett** [comment]
The `What are those Weird Task Tags about?` link in getting started is broken at the moment. URL: https://exercism.org/docs/building/tasks

**Comment by BethanyG** [comment]
@DjangoFett - Thanks for the head's up.  😄   These are fixed in this now, and we'll update the other issues as we get time.

**Comment by BethanyG** [comment]
Closing for rewrite.
